<p style="text-align:center"> 
<a href="https://skills.network" target="_blank"> 
<img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo"> 
</a>
</p>

# <h1 align="center"><font size="7"><strong>Final project</strong></font></h1>
## <h2 align= "center"><font size="6.8">**Homes for sale in King County, USA**</font></h2>

<hr>

## Part 2: Data Manipulation

In [1]:
# We import the libraries
import pandas as pd
import numpy as np

import warnings
warnings.filterwarnings('ignore')

Once we have our data, we can begin processing it.

The first step is to normalize the data; that is, decide whether we need to clean, complete, or format columns and/or values ​​that might affect our analysis.

### Data Format
As we saw with the ``describe()`` method, we have the only column, 'date,' in ``object`` format, that is, as a categorical variable composed of strings.
We must format it to a numeric data type (int or float) so that it doesn't affect the analysis. For dates, we use the `datetime` data type.

In [2]:
df = pd.read_csv('../data/raw/data.csv')

In [3]:
# Format the 'date' column to datetime format
df['date'] = pd.to_datetime(df['date'], format='%Y%m%dT%H%M%S')

# Verify that the conversion has been completed
print(f'Formatted date from object to: {df["date"].dtype}')

Formatted date from object to: datetime64[ns]


We've confirmed that the `describe()` method now takes into account the ``date`` column. This method only considers numeric columns (int, float, or datetime).

In [4]:
df.describe()

Unnamed: 0.1   Unnamed: 0            id                           date  \
count   21613.00000  21613.00000  2.161300e+04                          21613   
mean    10806.00000  10806.00000  4.580302e+09  2014-10-29 04:38:01.959931648   
min         0.00000      0.00000  1.000102e+06            2014-05-02 00:00:00   
25%      5403.00000   5403.00000  2.123049e+09            2014-07-22 00:00:00   
50%     10806.00000  10806.00000  3.904930e+09            2014-10-16 00:00:00   
75%     16209.00000  16209.00000  7.308900e+09            2015-02-17 00:00:00   
max     21612.00000  21612.00000  9.900000e+09            2015-05-27 00:00:00   
std      6239.28002   6239.28002  2.876566e+09                            NaN   

              price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21600.000000  21603.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.372870      2.115736   2079.899736  1.510697e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   
std    3.671272e+05      0.926657      0.768996    918.440897  4.142051e+04   

             floors  ...         grade    sqft_above  sqft_basement  \
count  21613.000000  ...  21613.000000  21613.000000   21613.000000   
mean       1.494309  ...      7.656873   1788.390691     291.509045   
min        1.000000  ...      1.000000    290.000000       0.000000   
25%        1.000000  ...      7.000000   1190.000000       0.000000   
50%        1.500000  ...      7.000000   1560.000000       0.000000   
75%        2.000000  ...      8.000000   2210.000000     560.000000   
max        3.500000  ...     13.000000   9410.000000    4820.000000   
std        0.539989  ...      1.175459    828.090978     442.575043   

           yr_built  yr_renovated       zipcode           lat          long  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean    1971.005136     84.402258  98077.939805     47.560053   -122.213896   
min     1900.000000      0.000000  98001.000000     47.155900   -122.519000   
25%     1951.000000      0.000000  98033.000000     47.471000   -122.328000   
50%     1975.000000      0.000000  98065.000000     47.571800   -122.230000   
75%     1997.000000      0.000000  98118.000000     47.678000   -122.125000   
max     2015.000000   2015.000000  98199.000000     47.777600   -121.315000   
std       29.373411    401.679240     53.505026      0.138564      0.140828   

       sqft_living15     sqft_lot15  
count   21613.000000   21613.000000  
mean     1986.552492   12768.455652  
min       399.000000     651.000000  
25%      1490.000000    5100.000000  
50%      1840.000000    7620.000000  
75%      2360.000000   10083.000000  
max      6210.000000  871200.000000  
std       685.391304   27304.179631  

[8 rows x 23 columns]

### Remove Unnecessary Data
Let's identify and remove unnecessary columns or rows, or at least those that don't directly influence the data in the dataset, such as in this case <code>"id"</code> and <code>"Unnamed: 0"</code> or <code>"date"</code>, which specifies the day the house was sold.

> **Note:**
> Although we could have skipped the first step, it was decided to do it anyway as a practical demonstration of how to proceed in situations where this variable needs to be formatted.

In [ ]:
# We remove irrelevant data, in this case the 'id', 'Unnamed: 0.1', 'Unnamed: 0' and 'date' columns.
try:
    df.drop(columns='id', inplace=True)
    df.drop(columns='Unnamed: 0.1', inplace=True)
    df.drop(columns='Unnamed: 0', inplace=True)
    df.drop(columns='date', inplace=True)
    print('The columns have been deleted.')
except KeyError:
    print('The columns have already been deleted.')

The columns have been deleted.


In [6]:
# We verify that the columns were deleted correctly
df.head()

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0  221900.0       3.0       1.00         1180      5650     1.0           0   
1  538000.0       3.0       2.25         2570      7242     2.0           0   
2  180000.0       2.0       1.00          770     10000     1.0           0   
3  604000.0       4.0       3.00         1960      5000     1.0           0   
4  510000.0       3.0       2.00         1680      8080     1.0           0   

   view  condition  grade  sqft_above  sqft_basement  yr_built  yr_renovated  \
0     0          3      7        1180              0      1955             0   
1     0          3      7        2170            400      1951          1991   
2     0          3      6         770              0      1933             0   
3     0          5      7        1050            910      1965             0   
4     0          3      8        1680              0      1987             0   

   zipcode      lat     long  sqft_living15  sqft_lot15  
0    98178  47.5112 -122.257           1340        5650  
1    98125  47.7210 -122.319           1690        7639  
2    98028  47.7379 -122.233           2720        8062  
3    98136  47.5208 -122.393           1360        5000  
4    98074  47.6168 -122.045           1800        7503

We generate a statistical summary again, but now without the columns we discarded in the previous step.

In [7]:
df.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.161300e+04  21600.000000  21603.000000  21613.000000  2.161300e+04   
mean   5.400881e+05      3.372870      2.115736   2079.899736  1.510697e+04   
std    3.671272e+05      0.926657      0.768996    918.440897  4.142051e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.219500e+05      3.000000      1.750000   1427.000000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.618000e+03   
75%    6.450000e+05      4.000000      2.500000   2550.000000  1.068800e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21613.000000  21613.000000  21613.000000  21613.000000  21613.000000   
mean       1.494309      0.007542      0.234303      3.409430      7.656873   
std        0.539989      0.086517      0.766318      0.650743      1.175459   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21613.000000   21613.000000  21613.000000  21613.000000  21613.000000   
mean    1788.390691     291.509045   1971.005136     84.402258  98077.939805   
std      828.090978     442.575043     29.373411    401.679240     53.505026   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1951.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4820.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000   21613.000000   21613.000000  
mean      47.560053   -122.213896    1986.552492   12768.455652  
std        0.138564      0.140828     685.391304   27304.179631  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471000   -122.328000    1490.000000    5100.000000  
50%       47.571800   -122.230000    1840.000000    7620.000000  
75%       47.678000   -122.125000    2360.000000   10083.000000  
max       47.777600   -121.315000    6210.000000  871200.000000

### Identifying NaN, Null, or Missing Values
Now we look for missing values. We can see that we have missing values ​​for the <code> bedrooms</code> and <code> bathrooms</code> columns by looking at the `count` row, where we have 21,600 and 21,603 respectively, instead of the total number of rows in the dataset, which is 21,613.

In [8]:
print("Number of NaN values ​​for the bedrooms column:", df['bedrooms'].isnull().sum())
print("Number of NaN values ​​for the bathrooms column:", df['bathrooms'].isnull().sum())

Number of NaN values ​​for the bedrooms column: 13
Number of NaN values ​​for the bathrooms column: 10


Missing values ​​were identified in our dataset. Various statistical strategies exist to address this problem, such as imputation using the mean of the data or the elimination of incomplete records when their value cannot be reasonably estimated. In this case, we chose to replace the missing values ​​using imputation techniques, with the goal of preserving as much information as possible and not reducing the size of the dataset.

In [9]:
# Replace NaN values ​​with the column average
mean = df['bedrooms'].mean()
df['bedrooms'].replace(np.nan, mean, inplace=True)
print('The data was replaced successfully.')

The data was replaced successfully.


We also replace missing values ​​in the ``'bathrooms'`` column with the mean of the ``'bathrooms'`` column itself.

In [10]:
# Replace NaN values ​​with the column average
mean=df['bathrooms'].mean()
df['bathrooms'].replace(np.nan,mean, inplace=True)
print('The data was replaced successfully.')

The data was replaced successfully.


We check again if we have missing values ​​in our data frame

In [11]:
print("Number of NaN values ​​for the bedrooms column:", df['bedrooms'].isnull().sum())
print("Number of NaN values ​​for the bathrooms column:", df['bathrooms'].isnull().sum())

Number of NaN values ​​for the bedrooms column: 0
Number of NaN values ​​for the bathrooms column: 0


### Identifying Outliers
First, an outlier is an observation in a data set that is significantly different from the rest of the data. These values ​​can be the result of measurement errors, incorrectly entered data, extreme phenomena, or legitimate but rare cases. In this case, we have extreme values, but they are clearly real, since when estimating the price of a house, whether it faces the sea or not will be a clear price trigger. We confirm this by observing that it is the column that generates the largest ranges of values.

Therefore, we will consider determining the IQR by taking into account the upper quartile limit of houses with sea views (df['waterfront']).

``Any value outside this range is considered an outlier.``

In [12]:
# Filter the DataFrame to get only the rows where waterfront = 1
df_waterfront = df[df['waterfront'] == 1]

# Calculate the upper and lower bounds to detect outliers
Q1 = df_waterfront.quantile(0.25)
Q3 = df_waterfront.quantile(0.75)
IQR = Q3 - Q1

# Filter outliers
df_without_atypicals = df_waterfront[~((df_waterfront < (Q1 - 1.5 * IQR)) | (df_waterfront > (Q3 + 1.5 * IQR))).any(axis=1)]

# If you want to keep the original DataFrame without the outliers where waterfront = 1
df_updated = pd.concat([df[df['waterfront'] != 1], df_without_atypicals])

print(f'The original DataFrame had {df.shape[0]} rows and now has {df_updated.shape[0]} rows.')

The original DataFrame had 21613 rows and now has 21558 rows.


We apply the describe method again to see how the descriptive statistics of the dataset turned out.

In [13]:
df_updated.describe()

price      bedrooms     bathrooms   sqft_living      sqft_lot  \
count  2.155800e+04  21558.000000  21558.000000  21558.000000  2.155800e+04   
mean   5.369100e+05      3.373451      2.114396   2076.573755  1.503397e+04   
std    3.537667e+05      0.925113      0.766176    910.236773  4.130076e+04   
min    7.500000e+04      1.000000      0.500000    290.000000  5.200000e+02   
25%    3.210000e+05      3.000000      1.750000   1420.500000  5.040000e+03   
50%    4.500000e+05      3.000000      2.250000   1910.000000  7.605500e+03   
75%    6.420000e+05      4.000000      2.500000   2550.000000  1.063700e+04   
max    7.700000e+06     33.000000      8.000000  13540.000000  1.651359e+06   

             floors    waterfront          view     condition         grade  \
count  21558.000000  21558.000000  21558.000000  21558.000000  21558.000000   
mean       1.494155      0.005010      0.226459      3.408943      7.654374   
std        0.539967      0.070604      0.750328      0.650104      1.171038   
min        1.000000      0.000000      0.000000      1.000000      1.000000   
25%        1.000000      0.000000      0.000000      3.000000      7.000000   
50%        1.500000      0.000000      0.000000      3.000000      7.000000   
75%        2.000000      0.000000      0.000000      4.000000      8.000000   
max        3.500000      1.000000      4.000000      5.000000     13.000000   

         sqft_above  sqft_basement      yr_built  yr_renovated       zipcode  \
count  21558.000000   21558.000000  21558.000000  21558.000000  21558.000000   
mean    1786.097041     290.476714   1971.046386     83.509741  98077.921143   
std      823.591225     439.837084     29.366964    399.650901     53.506782   
min      290.000000       0.000000   1900.000000      0.000000  98001.000000   
25%     1190.000000       0.000000   1952.000000      0.000000  98033.000000   
50%     1560.000000       0.000000   1975.000000      0.000000  98065.000000   
75%     2210.000000     560.000000   1997.000000      0.000000  98118.000000   
max     9410.000000    4130.000000   2015.000000   2015.000000  98199.000000   

                lat          long  sqft_living15     sqft_lot15  
count  21558.000000  21558.000000   21558.000000   21558.000000  
mean      47.560185   -122.213625    1984.830504   12711.132155  
std        0.138595      0.140748     683.522887   27262.865849  
min       47.155900   -122.519000     399.000000     651.000000  
25%       47.471300   -122.328000    1490.000000    5100.000000  
50%       47.571900   -122.230000    1840.000000    7620.000000  
75%       47.678100   -122.125000    2360.000000   10063.750000  
max       47.777600   -121.315000    6210.000000  871200.000000

We see how the average value (mean) of the price column is no longer ``5.400881e+05`` but ``5.360473e+05``. Since we limited the range in the total data set by dividing it into two groups, taking the group with the largest data range and discarding the outliers from the latter. All of this to have the largest possible data set to be able to train our future model.

In [14]:
# Save the dataset again, but with the normalized data
df_updated.to_csv('../data/processed/data_clean.csv')

<hr>

## Author

<a href="https://www.linkedin.com/in/flavio-aguirre-12784a252/">**Flavio Aguirre**</a><br>
<a href="https://coursera.org/share/e27ae5af81b56f99a2aa85289b7cdd04">***Data Scientist***</a>